(request :post "url" params) doesn't include the params

aredington opened the pull request:
While doing some Clojure web development, Tim Ewald and I have had some complications using ring-mock through kerodon. After tracing the problem, we think this patch will resolve our issues. The changes in this request:
- Only feed params into query-string if the verb is get
- Feed params into the body if the verb is post
- Move some fns earlier in the file so request can use them
- Switch body from a multimethod to a protocol

We could remove the protocol impl if that's important, but felt it would be a good overall change and would include it since we were messing with body already. We updated tests to reflect the new return values of request when the verb is post.

Commit:
Prevent request from clobbering parameters

- Only feed params into query-string if the verb is get
- Feed params into the body if the verb is post
- Move some fns earlier in the file so request can use them
- Switch body from a multimethod to a protocol

weavejester commented:
I don't really understand the purpose of the protocol change. Could you explain the benefit?

The proposed change also removes the functionality to set query parameters on POSTs. I agree this probably shouldn't be the default, but I think we need an alternative mechanism, perhaps by having a `query-string` function that alters a request, e.g.

``` clojure
(request :get "/" {:q "foo"})   ;; shortcut

(-> (request :get "/")
    (query-string {:q "foo"}))  ;; long way

(-> (request :post "/")
    (query-string {:q "bar"}))  ;; also works with POST

(request :post "/" {:q "bar"})) ;; params in POST body instead

(-> (request :post "/")
    (body {:q "bar"}))          ;; equivalent to the above shortcut
```

neatonk committed:
Handle request params differently for :get and :post methods.

As a shortcut, params provided for a :get will be appended to the
:query-string and params provided for a :post will be used as the
request :body.

* moved request down to make 'body and 'query-string available
* added new 'query-string fn for modifying a request.
* preserves existing :query-string when using 'query-string.
* replace old 'query-string with simpler 'append-query.
* added test cases for 'adding params in :post'

Commit:
Merge branch 'master' of https://github.com/neatonk/ring-mock

Conflicts:
	src/ring/mock/request.clj

Commit:
Make query-string clobber the query string (consistent with body)

aredington commented:
@neatonk was kind enough to implement your feedback! I've merged his work into my fork and added a test around query-string's semantics after discussing it with him.

@timewald and I both felt that switching from a multimethod to a protocol made the dispatch by type somewhat more evident and that was the real reason that we made the change. Aside from that there is the claim that protocols perform better than multimethods. I tried running both the mutlimethod dispatch and protocol dispatch side by side in the same pre-heated jvm; my methodology was to run 100000 trials of invoking each against a constant request body, 10 times under clojure.core/time, and to average the results. My results suggest that using the protocol is 5% faster than multimethods. That's not a huge win, so if you'd like me to revert my fork to the multimethods I'd be happy to.

weavejester commented:
I'm not too concerned about speed when it comes to testing, especially if there's only a 5% difference. Could you remove the protocol changes and then tidy up the commits a little (e.g. remove any merge commits, make sure the commits are generally neat, etc.)

aredington commented:
@weavejester Some quick googling suggests there's no way to alter which branch a pull request points at. I've created a branch from 7ba9e4 and created a series of commits that I hope are more to your liking at https://github.com/aredington/ring-mock/tree/clean-history. This branch also preserves the multi-method implementation of body. If you like I can close this pull request and reopen against that branch, or you can pull the changes in without GitHub process involved.

weavejester commented:
You can't change the branch at which a pull request points to, but you can change the commits in a branch pretty easily. For instance, you could do:

```
git checkout clean-history
git push -f origin master
```

The `-f` argument forces an update that might result in existing commits being changed.

That all said, I'll probably merge things in manually this time around :)
